agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1504 @@
+    /// /// or an error being returned from the discovery handler (that is not a broken pipe)
+    /// /// DiscoveryHandlerConnectivityStatus::Offline if a connection cannot be established via a call to get_stream
+    /// If a connection cannot be established, continues to try, sleeping between iteration.
+    /// Removes the discovery handler from the RegisteredDiscoveryHandlerMap if it has been offline for longer than the grace period.
+    pub async fn do_discover(
+        discovery_operator: Arc<DiscoveryOperator>,
+        kube_interface: Arc<Box<dyn k8s::KubeInterface>>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let config = discovery_operator.get_config();
+        trace!(
+            "do_discover - entered for protocol {}",
+            config.spec.protocol.name
+        );
+        // get clone of map
+        let mut discovery_handler_map = discovery_operator
+            .get_discovery_handler_map()
+            .lock()
+            .unwrap()
+            .clone();
+        trace!(
+            "do_discover - discovery_handler_map is {:?}",
+            discovery_handler_map
+        );
+        if let Some(protocol_dhs_map) = discovery_handler_map.get_mut(&config.spec.protocol.name) {
+            for (endpoint, dh_details) in protocol_dhs_map.clone() {
+                trace!(
+                    "do_discover - for protocol {} and endpoint {:?}",
+                    config.spec.protocol.name,
+                    endpoint
+                );
+                // Only use Discovery Handler if it doesn't have a client yet
+                if dh_details.connectivity_status != DiscoveryHandlerConnectivityStatus::HasClient {
+                    trace!(
+                        "do_discover - endpoint {:?} for protocol {} doesn't have client",
+                        endpoint,
+                        config.spec.protocol.name
+                    );
+                    let mut stop_discovery_receiver = dh_details.stop_discovery.subscribe();
+                    loop {
+                        tokio::select! {
+                            _ = try_receive(&mut stop_discovery_receiver) => {
+                                trace!("do_discover - received message to stop discovery for discovery handler at endpoint {:?} for configuration {}", endpoint, discovery_operator.get_config().metadata.name);
+                                break;
+                            },
+                            _ = do_discover_on_discovery_handler(discovery_operator.clone(), kube_interface.clone(), &endpoint, &dh_details) => {
+                                trace!("do_discover - discovery completed for discovery handler at endpoint {:?} for configuration {}", endpoint, discovery_operator.get_config().metadata.name);
+                                break;
+                            }
+                        }
+                    }
+                }
```

Comment:
should there be some logging if this (for some reason) is called when HasClient?

Reply:
maybe this is a case we could see if someone deploys a DH daemonset twice accidentally/mistakenly?

Reply:
do_discover is called when a Configuration is first applied and upon being notified of any new discover handlers (of the same protocol type). So if one discovery handler was already being used (HasClient) and another registers, this is a supported scenario rather than an error. The `DiscoveryOperator` was build so as to support multiple DH (daemonsets) being applied. I didn't want to assume only 1 DH per protocol -- maybe someone wants HA of their Discovery Handler.